core/services/ocr2/plugins/s4/integration_test.go

```diff
@@ -0,0 +1,406 @@
+		err := don.orms[0].Update(row, pg.WithParentCtx(ctx))
+		require.NoError(t, err)
+	}
+	originSnapshot, err := don.orms[0].GetSnapshot(s4_svc.NewFullAddressRange(), pg.WithParentCtx(ctx))
+	require.NoError(t, err)
+
+	// It requires at least two rounds due to MaxObservationEntries = rows / 2
+	errors := don.executeOCR(ctx, 2)
+	checkNoErrors(t, errors)
+
+	for i := 1; i < don.size; i++ {
+		snapshot, err := don.orms[i].GetSnapshot(s4_svc.NewFullAddressRange(), pg.WithParentCtx(ctx))
+		require.NoError(t, err)
+		equal := compareSnapshots(originSnapshot, snapshot)
+		assert.True(t, equal, "oracle %d", i)
+	}
+}
+
+func TestS4Integration_Expired(t *testing.T) {
+	config := createPluginConfig(100)
+	config.MaxObservationEntries = 5
+	don := newDON(t, 4, config)
+	ctx := testutils.Context(t)
+
+	// injecting expiring records
+	rows := generateTestOrmRows(t, 10, time.Millisecond)
+	for _, row := range rows {
+		err := don.orms[0].Update(row, pg.WithParentCtx(ctx))
+		require.NoError(t, err)
+	}
+
+	// within one round, all records will be GC-ed
+	time.Sleep(testutils.TestInterval)
+	errors := don.executeOCR(ctx, 1)
+	checkNoErrors(t, errors)
+
+	for i := 0; i < don.size; i++ {
+		snapshot, err := don.orms[i].GetSnapshot(s4_svc.NewFullAddressRange(), pg.WithParentCtx(ctx))
+		require.NoError(t, err)
+		require.Len(t, snapshot, 0)
+	}
+}
+
+func TestS4Integration_NSnapshotShards(t *testing.T) {
+	config := createPluginConfig(10000)
+	config.NSnapshotShards = 4
+	don := newDON(t, 4, config)
+	ctx := testutils.Context(t)
+
+	// injecting lots of new records (to be close to normal address distribution)
+	rows := generateTestOrmRows(t, 1000, time.Minute)
+	for _, row := range rows {
+		err := don.orms[0].Update(row, pg.WithParentCtx(ctx))
+		require.NoError(t, err)
+	}
+	originSnapshot, err := don.orms[0].GetSnapshot(s4_svc.NewFullAddressRange(), pg.WithParentCtx(ctx))
+	require.NoError(t, err)
+
+	// this still requires one round, because Observation takes all unconfirmed rows
+	errors := don.executeOCR(ctx, 1)
+	checkNoErrors(t, errors)
+
+	for i := 1; i < don.size; i++ {
+		snapshot, err := don.orms[i].GetSnapshot(s4_svc.NewFullAddressRange(), pg.WithParentCtx(ctx))
+		require.NoError(t, err)
+		equal := compareSnapshots(originSnapshot, snapshot)
+		assert.True(t, equal, "oracle %d", i)
+		checkNoUnconfirmedRows(ctx, t, don.orms[i], 1000)
+	}
+}
+
+func TestS4Integration_OneNodeOutOfSync(t *testing.T) {
+	don := newDON(t, 4, createPluginConfig(100))
+	ctx := testutils.Context(t)
+
+	// injecting same confirmed records to all nodes but the last one
+	rows := generateConfirmedTestOrmRows(t, 10, time.Minute)
+	for o := 0; o < don.size-1; o++ {
+		for _, row := range rows {
+			err := don.orms[o].Update(row, pg.WithParentCtx(ctx))
+			require.NoError(t, err)
+		}
+	}
+
+	// all records will be propagated to the last node when it is a leader
+	// leader selection is round-robin, so the 4th iteration picks the last node
+	errors := don.executeOCR(ctx, 4)
```

Comment:
Awesome test framework! So easy to create OCR plugin tests :) I wonder if Research built something similar cc @stchrysa